### PR TITLE
Improve OrderPreservingMultiDictionary.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
1. Support allocation-free enumeration of the dictionary.
2. Pool internal data so that less garbage is churned when creating and releasing dictionaries.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide  @dotnet/roslyn-compiler 

The IDE would like to use this API to have a multidictionary that has low memory churn characteristics.  Problems with the current API were that:

1. there was no allocation-free way of enumerating the values within.  You could request "Keys" but that would allocate an object for the KeyCollection.  
2. Even though the dictionary exposed a pooling API, the pooling only went as far as the top level dictionary itself.  Internal data (specifically the arraybuilders used to store values) were not pooled and would cause GC churn.

This PR addresses both these concerns.

**Comment by CyrusNajmabadi** [comment]
test vsi please

File: src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs [hunk 1]
```diff
@@ -57,159 +65,237 @@ public void Free()
...
+            /// Each value is either a single V or an <see cref="ArrayBuilder{V}"/>.
+            /// Never null.
+            /// </summary>
+            private readonly object _value;
+
+            // By default we allocate array builders with a size of two.  That's two store
+            // the single item already in _value, and to store the item we're adding.  
+            // In general, we presume that the amount of values per key will be low, so this
+            // means we have very little overhead when there are multiple keys per value.
+            private static ObjectPool<ArrayBuilder<V>> s_builderPool = new ObjectPool<ArrayBuilder<V>>(
```
Comment (gafter) [inline src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs]: Isn't there a pooled version of array builder? e.g. `ArrayBuilder<V>.GetInstance()`

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs]: Nice catch!  I've switched to using that instead!

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
test vsi please

**Comment by CyrusNajmabadi** [comment]
retest roslyn_prtest_win_vsi0 please

File: src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs [hunk 2]
```diff
@@ -57,159 +65,231 @@ public void Free()
...
-        /// </summary>
-        public ImmutableArray<V> this[K k]
+        public struct ValueSet : IEnumerable<V>
         {
-            get
+            /// <summary>
+            /// Each value is either a single V or an <see cref="ArrayBuilder{V}"/>.
+            /// Never null.
+            /// </summary>
+            private readonly object _value;
```
Comment (cston) [inline src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs]: Is it common to have multiple items with the same key? If not, consider using ```OneOrMany<T>``` instead, which stores the values as `T` or ```ImmutableArray<T>```.

Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs]: Is that any better though?  In the current model, 1, 2, and many are handled fine.  In the OneOrMany case, the moment you go past 2, you're always going to allocate on each subsequent add.

**Comment by gafter** [comment]
LGTM, after the noted change to use an existing pool.

File: src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs [hunk 3]
```diff
@@ -57,159 +65,231 @@ public void Free()
...
+                        return arrayBuilder[index];
+                    }
+                }
+            }
+
+            internal ImmutableArray<V> Items
+            {
+                get
+                {
+                    Debug.Assert(this.Count >= 1);
```
Comment (VSadov) [inline src/Compilers/Core/Portable/Collections/OrderPreservingMultiDictionary.cs]: Is it common to fetch same "Items" twice? 
Perhaps it is worth storing the array once it is created. (array can be dropped when set is modified).

**Comment by VSadov** [comment]
LGTM. 
Consider caching the array created in "Items". It could save some more allocation if re-reading same arrays is common enough

**Pull request merged**